File: scala/algorithms/warmup/DiagonalDifference.scala
package algorithms.warmup

// In order to submit to hackerrank, rename it to "Solution"
object DiagonalDifference {
  def main(args: Array[String]) = {
    val scanner = new java.util.Scanner(System.in)

    val dimension = scanner.nextLine.toInt
    val matrixValuesFromInput = (0 until dimension).map(d => scanner.nextLine.split(" ").map(_.toInt))

    val matrix = Matrix(dimension, matrixValuesFromInput)
    val diagonalDifference = (matrix.sumAcrossThePrimaryDiagonal - matrix.sumAcrossTheSecondaryDiagonal).abs

    println(diagonalDifference)
  }
}

case class Matrix(dimension: Int, multiDimensionArray: Seq[Array[Int]]) {
  def sumAcrossThePrimaryDiagonal: Int = {
    (0 until dimension).map(d => multiDimensionArray(d)(d)).sum
  }

  def sumAcrossTheSecondaryDiagonal: Int = {
    (0 until dimension).map(d => multiDimensionArray((dimension - 1) -  d)(d)).sum
  }
}

File: c#/algorithms/warmup/DiagonalDifference.cs
using System;
using System.Linq;
class Solution {

  private static int[][] BuildMatrix(int matrixDimension)
  {
    return Enumerable.Range(0, matrixDimension)
      .Select(_ => Console.ReadLine().Split(' ').Select(n => Convert.ToInt32(n)).ToArray())
      .ToArray();
  }

  private static int SumAcrossThePrimaryDiagonal(int[][] matrix)
  {
    return Enumerable.Range(0, matrix.Length)
      .Select(d => matrix[d][d])
      .Aggregate((curr, next) => curr + next);
  }

  private static int SumAcrossTheSecondaryDiagonal(int[][] matrix)
  {
    return Enumerable.Range(0, matrix.Length)
      .Select(d => matrix[d][(matrix.Length-1) - d])
      .Aggregate((curr, next) => curr + next);
  }

  static void Main(String[] args)
  {
    int matrixDimension = Convert.ToInt32(Console.ReadLine());
    int[][] matrix = BuildMatrix(matrixDimension);

    int diagonalDifference = Math.Abs(
      SumAcrossThePrimaryDiagonal(matrix) - SumAcrossTheSecondaryDiagonal(matrix)
    );

    Console.WriteLine(diagonalDifference);
  }
}
